crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,531 @@
+/// line.
+type NameDeviceMapping = HashMap<&'static str, DeviceSpecification>;
+
+/// Creates device mapping and adds all the supported devices to it.
+///
+/// If you want to introduce a new kind of host device bindings, insert a
+/// device identifier + [DeviceSpecification] to [NameDeviceMapping] here.
+pub fn available_devices() -> NameDeviceMapping {
+    let mut devices: NameDeviceMapping = HashMap::new();
+    devices.insert(
+        "fake",
+        DeviceSpecification::new(
+            fake::FakeSensor::new_device,
+            "fake accelerometer",
+            "A simple 3-axes sensor providing fake pre-defined values.",
+            vec!["name: an optional name of the sensor, max. 15 characters"],
+        ),
+    );
+    devices.insert(
+        "iio",
+        DeviceSpecification::new(
+            iio::IIOSensor::new_device,
+            "industrial I/O sensor",
+            "",
+            vec![
+                "path: path to the device directory (e.g. /sys/bus/iio/devices/iio:device0)",
+                "channel: prefix of the device type (e.g. in_accel)",
+                "name: an optional name of the sensor, max. 15 characters",
+            ],
+        ),
+    );
+    devices
+}
+
+fn devices_help() -> String {
+    let mut help = String::new();
+    writeln!(help, "Available devices:").unwrap();
+    for (name, specification) in available_devices().iter() {
+        let short_help = &specification.short_help;
+        let long_help = &specification.long_help;
+        let parameters_help = &specification.parameters_help;
+        writeln!(help, "\n- {name}: {short_help}").unwrap();
+        for line in long_help.lines() {
+            writeln!(help, "  {line}").unwrap();
+        }
+        if !parameters_help.is_empty() {
+            writeln!(help, "  Parameters:").unwrap();
+            for parameter in parameters_help {
+                writeln!(help, "  - {parameter}").unwrap();
+            }
+        }
+    }
+    writeln!(help, "\nDevice specification example:").unwrap();
+    writeln!(
+        help,
+        "--device iio,path=/sys/bus/iio/devices/iio:device0,channel=in_accel"
+    )
+    .unwrap();
+    help
+}
+
+pub fn print_devices_help() {
+    let help = devices_help();
+    println!("{}", help);
+}
+
+pub fn check_device_properties(
+    properties: &DeviceProperties,
+    required: &[&str],
+    optional: &[&str],
+) -> Result<(), String> {
+    let missing = properties.missing(required);
+    if !missing.is_empty() {
+        return Err(format!(
+            "Missing device parameters: {}",
+            missing.iter().sorted().join(", ")
+        ));
+    }
+    let mut all_allowed = Vec::from(required);
+    all_allowed.extend(optional.iter());
+    let extra = properties.extra(&all_allowed);
+    if !extra.is_empty() {
+        return Err(format!(
+            "Unexpected device parameters: {}",
+            extra.iter().sorted().join(", ")
+        ));
+    }
+    Ok(())
```

Comment:
```suggestion
    if !extra.is_empty() {
        Err(format!(
            "Unexpected device parameters: {}",
            extra.iter().sorted().join(", ")
        ))
    } else {
        Ok(())
    }
```

Reply:
I don't think it would be a good idea to use different patterns for `missing' and `extra' in the function.